### ISSUE TITLE
Enforce pagination on catalog entity requests

### ISSUE THREAD

**Issue by tkrop** [backstage/backstage#5921]
Pagination is an important goals to reach two goals:

1. Protect a service against overload.
2. Improve user experience on requests.

Currently, pagination on catalog entities (implemented in https://github.com/backstage/backstage/pull/4993) is optional targeting the 2nd goal. However, this strategy imposes the risk, that plugin contributors accidentally implement request that overload the backstage backend.

## Feature Suggestion

To protect the backend, the catalog API must enforce pagination on all requests, that might overload the backend resources.

## Possible Implementation

 While it may be possible to determine potentially risky requests, this creates complexities and efforts, that I think is not worth pursuing this approach. Instead I suggest to just enforce pagination for all requests based on a configurable default limit.

Since enforcing pagination is a breaking change, I suggest to make enforced pagination configurable (default: to `true`) to give the community time to disable enforcing and adjust to this change.

## Context

In our system we observed a resource overload, that made a backend too unresponsive for the Kubernetes health check, that consequently restarted the backend making the situation worse. The situation was triggered by an incorrect filter configuration triggering a request very similar too:

```
curl 'https://<server>/catalog/entities?filter='
```

**Comment by freben** [comment]
I'd be up for making the backend have a `maxPageSize?: number` config parameter.

From the caller side, the `catalog-client` could use forced pagination internally with a fixed but fairly large page size, for naked `entities()` calls, if we wanted to make it harder to make unintentional mistakes.

**Comment by tkrop** [comment]
Sounds like an alias for `defaultLimit`, so I expect it will solve the same problem - if there is a way to set it practically to unlimited for backward compatibility.

@freben by the way, may be you like to have a look on our API guideline, how we implement pagination in Zalando usually?

Regarding `catalog-client`: I would take the term protection against overload here very serious. We trust our contributors, but they evidently do mistakes. Consequently, I suggest to not support naked `entities()` requests, but always rely on pagination-based interfaces.

**Comment by freben** [comment]
Right, but if catalog-client has absolutely no way for the caller to make a non-paged request (if you don't ask for paging it'll do it transparently under the hood anyway and return the union of it all when done), and the server refuses to serve requests that don't have (valid) paging, then we should be good right?

**Comment by tkrop** [comment]
@freben Yes, exactly - will work fine! For me the second part `server refuses to serve requests` is important. I was just considering, that some backstage users may by-pass the `catalog-client` and thereby experience a breaking change, when the rejection behavior cannot be switched on and off, or more precisely isn't switched of by default.

Any enforcing solution is fine for us.

**Comment by Rugvip** [comment]
Typical thing to do here is to have an implicit and potentially capped `limit` query right? So if you send a request for listing all entities without any additional parameters, you'll just get the first page back along with the actual limit that you ended up getting. If you try to set the `limit` above the configured maximum, you simply won't get that many items back and the response will tell you how many you actually got.

**Comment by tkrop** [comment]
Yes, I think this would be the way to go: providing a very high default limit or even `0` for infinite.

Do you think there is a way to detect plugins, that side-step the client and need to made aware of pagination? It could be something like adding a limit to the client and warn in the log on request without limit. Breaking the request to create awareness would even be better, if the number of entities required to be delivered is higher than the limit.

**Comment by recobe182** [comment]
With enforcing pagination, I assume the behavior of Owned, Starred and All filters in `UserListPicker` will change? as it doesn't have a whole dataset to work on. I found it useful to know the total number of entries in the catalog.

**Comment by freben** [comment]
Yes, that's one of the complications.

To be able to show match-count dots you either have to have backend support for explicitly resolving that (which we don't have in place), or to fetch the entire result set (possibly page by page, but with less benefit than probably initially hoped for).

Note that some filters are server side, and some are client side post-filters (such as favorites, which the catalog has no notion of).

Also the table implementation we use prefers to have the entire dataset upfront to be able to drive the visual pagination controls. It would probably need some code tweaks to handle this better.

That all being said, enforcing paging probably can be a wise thing nonetheless, forcing the caller to fetch several smaller chunks, which breaks work into smaller atoms and doesn't as easily overwhelm the server.

**Comment by kristina-hager** [comment]
I'm interested in this topic and fix. Has anyone started working on code for this? Do you have any guidance on how to proceed here if me/my team is interested in fixing this issue? We have 7K items in our catalog and growing, so this is an issue for us.

**Comment by freben** [comment]
@kristina-hager Hi! Not that we know of. There is basic paging support already on the backend side, but as the issue describes, there may be more work to do to make it more fleshed out end-to-end. We'd love contributions in this space. Do you have an idea so far what the issues are that you see, and what needs you have?

**Comment by dehamzah** [comment]
I also facing the same problem, we had about 40k entities, and it's really painful to the server if someone calls `entityClient.getEntities()` without the filter. At the moment, there is also a plugin search that running periodically to index the catalog entities and tech docs.

Is there already a decision on how it will be enhanced on the catalog backend side?

**Comment by Rugvip** [comment]
I created #9139 as part of the way forward here. I think in the end it's gonna be a requirement for this as having the client run through all pages to form the catalog index page filters is likely still gonna be very slow and possibly load the catalog even more.

**Comment by Rugvip** [comment]
Couple of ideas on how the pagination can be tweaked over in #9905

**Comment by kunickiaj** [comment]
bump on this -- seems to still be an issue and makes me nervous as our `/api/catalog/entities` endpoint without any filters/pagination will return ~100MB of data...
